src/device_manager.rs

```diff
@@ -0,0 +1,431 @@
+
+//! System device management.
+//!
+//! [DeviceManager](struct.DeviceManager.html) responds to manage all devices
+//! of virtual machine, store basic device information like name and
+//! parent bus, register IO resources callback, unregister devices and help
+//! VM IO exit handling.
+
+extern crate vm_allocator;
+
+use self::vm_allocator::{Error as AllocatorError, SystemAllocator};
+use crate::device::*;
+use std::cmp::{Ord, Ordering, PartialEq, PartialOrd};
+use std::collections::btree_map::BTreeMap;
+use std::collections::HashMap;
+use std::result;
+use std::sync::{Arc, Mutex};
+use vm_memory::{Address, GuestAddress, GuestUsize};
+
+/// Guest physical address and size pair to describe a range.
+#[derive(Eq, Debug, Copy, Clone)]
+pub struct Range(pub GuestAddress, pub GuestUsize);
+
+impl PartialEq for Range {
+    fn eq(&self, other: &Range) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl Ord for Range {
+    fn cmp(&self, other: &Range) -> Ordering {
+        self.0.cmp(&other.0)
+    }
+}
+
+impl PartialOrd for Range {
+    fn partial_cmp(&self, other: &Range) -> Option<Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+/// Error type for `DeviceManager` usage.
+#[derive(Debug)]
+pub enum Error {
+    /// The insertion failed because the new device overlapped with an old device.
+    Overlap,
+    /// PIO request is none.
+    NonePIORequest,
+    /// The insertion failed because device already exists.
+    Exist,
+    /// The removing fails because the device doesn't exist.
+    NonExist,
+    /// Irq allocation failed.
+    IrqAllocate(AllocatorError),
+    /// Instance id allocation failed.
+    InstanceIdAllocate(AllocatorError),
+    /// Address allocation failed.
+    AddressAllocate(AllocatorError),
+}
+
+/// Simplify the `Result` type.
+pub type Result<T> = result::Result<T, Error>;
+
+/// System device manager serving for all devices management and VM exit handling.
+pub struct DeviceManager {
+    /// System allocator reference.
+    resource: Arc<Mutex<SystemAllocator>>,
+    /// Devices information mapped by instance id.
+    devices: HashMap<u32, DeviceDescriptor>,
+    /// Range mapping for VM exit mmio operations.
+    mmio_bus: BTreeMap<Range, Arc<Mutex<dyn Device>>>,
+    /// Range mapping for VM exit pio operations.
+    pio_bus: BTreeMap<Range, Arc<Mutex<dyn Device>>>,
+}
+
+impl DeviceManager {
+    /// Create a new `DeviceManager` with a `SystemAllocator` reference which would be
+    /// used to allocate resource for devices.
+    pub fn new(resource: Arc<Mutex<SystemAllocator>>) -> Self {
+        DeviceManager {
+            resource,
+            devices: HashMap::new(),
+            mmio_bus: BTreeMap::new(),
+            pio_bus: BTreeMap::new(),
+        }
+    }
+
+    fn insert(&mut self, dev: DeviceDescriptor) -> Result<()> {
+        // Insert if the key is non-present, else report error.
+        if self.devices.get(&(dev.instance_id)).is_some() {
```

Comment:
BtreeMap::contains_key() is designed for this:)

Reply:
Got it , thanks!